daemon/list.go

```diff
@@ -90,6 +90,41 @@ func (daemon *Daemon) Containers(config *types.ContainerListOptions) ([]*types.C
 	return daemon.reduceContainers(config, daemon.transformContainer)
 }
 
+func (daemon *Daemon) filterByNameIDMatches(ctx *listContext) []*container.Container {
+	names := ctx.filters.Get("name")
+	ids := ctx.filters.Get("id")
+	if len(names)+len(ids) == 0 {
+		// if name or ID filters are not in use, return to
+		// standard behavior of walking the entire container
+		// list from the daemon's in-memory store
+		return daemon.List()
+	}
+
+	matches := make(map[string]bool)
+	// look for name matches; errors only report not found and can be ignored
+	if len(names) > 0 {
+		for id, idNames := range ctx.names {
+			for _, eachName := range idNames {
+				if ctx.filters.Match("name", eachName) {
+					matches[id] = true
+				}
+			}
+		}
+	}
+	// look for ID matches; errors are not found and can be ignored
```

Comment:
It is a bit confusing as my first impl. for both loops (on ids and names) was to loop over the name or ID search terms; but because for names it can be a partial (regex-like) match anywhere in the name, it makes more sense to use the built-in matching while looping over the existing name map that is in the search context rather than write a "new impl." of the same thing.  However, to get into that search, we have to verify there are actually names to search.  For IDs, we are actually going to loop over the explicit search list so we don't need a "len > 0" check to enter that loop because as noted, if zero length, nothing will happen anyway.